#160577444 rating an article

koechkevin opened the pull request:
## What does this PR do?
 - Ensures that a user is able to rate an article with any rate in scale of 1 - 5
## Description of Task to be completed?
- 
## How should this be manually tested?
- git clone the repository and checkout to the branch
```https://github.com/andela/ah-shakas-frontend.git```
```cd ah-shakas-frontend```
```git checkout ft-rate-articles-160577444```
```yarn install```
- Create a `.env` file with the following values
  ```
  REACT_APP_BASE_URL='https://ah-shakas-staging.herokuapp.com/api'
  REACT_APP_UPLOAD_BASE_URL=
  REACT_APP_CLOUD_NAME=
  REACT_APP_UPLOAD_PRESET=
  ```
  The `CLOUD_NAME` and the `UPLOAD_PRESET` are obtained from your cloudinary account.

```yarn start```
- navigate through the articles to see the ratings of each article. You can login to make ratings on the article

## Any background context you want to provide?
- NA
## What are the relevant pivotal tracker stories?
[#160577444](https://www.pivotaltracker.com/story/show/160577444)

## Packages installed
- axios
- react-star-ratings
- redux-mock-store

## Deployment note
N/A

## Related PRs branch | PR (branch_name) | (pr_link)
Branch | PR
------ | ------
other_pr_production | [link]()
other_pr_master | [link]()

## Todos
- [x] Raise PR
- [ x] Test



## Screenshots (if appropriate)
<img width="950" alt="screen shot 2018-11-16 at 21 28 57" src="https://user-images.githubusercontent.com/39160236/48647712-d5564a80-e9fd-11e8-8c62-734ebec709d1.png">


## Questions:
<to be deleted>

koechkevin committed:
feat(rating): rate an article

- ensure stars component is created and rendered
- fetch ratings of particular article and display
- add tests for reducers and actions
- ensures all modules are tested
- use single quotes instead of double quotes
- add missing trailing comma dangles
- Ensure naming conventions are conformed
- Ensure test coverage is improved
- ensure documentation comments are placed appropriately
- separate average rating component and my rating component
- keep branch up to date with develop
- implement feedback as requested
- keep branch up to date with develop
- fix failing test
- deprecate mock article used in demo to actual article

[WIP #160577444]

Arthur236 commented on src/store/index.js:
```diff
@@ -1,16 +1,19 @@
 });
 
 const storeConfig = () => {
   const store = createStore(
```
Let's have each parameter on its own line. i.e;

```
const store = createStore(
    rootReducer,
    {},
    composeEnhancers(applyMiddleware(thunkMiddleware))
  );
```

koechkevin replied:
cool

Arthur236 commented on .eslintrc.json:
```diff
@@ -21,6 +21,7 @@
     "rules": {
-      "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }]
+      "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }],
+      "function-paren-newline": ["error", "never"]
```
This rule is not necessary. The default airbnb standard is what we should use.

Arthur236 commented on src/App.test.js:
```diff
@@ -6,6 +6,6 @@ jest.mock('react-router-dom');
 it('renders without crashing', () => {
   const div = document.createElement('div');
-  ReactDOM.render(<App />, div);
+  ReactDOM.render( < App / > , div);
```
Why the additional whitespaces here?

Arthur236 commented on src/Mock/mockArticle.jsx:
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
```
The mock folder should contain mock response data in the format that we expect it to come in from our api, not actual react components. Refactor this and rename the folder from `Mock` to `mock`

koechkevin replied:
noted.

Arthur236 commented:
Rename this file `src/__test__/reducers/Rating.test.js` to `src/__test__/reducers/rating.reducer.test.js`

Arthur236 commented:
Let's not have nested folders inside `actions` and `reducers` directories. Stick to the format and directory structure that we agreed on

Arthur236 commented:
Let's use describe - it blocks in our tests

Arthur236 commented on src/constants/index.js:
```diff
@@ -0,0 +1,5 @@
+export const RATE_ARTICLE = 'RATE_ARTICLE';
+export const AVERAGE_RATING = 'AVERAGE_RATING';
+export const RATING_ERROR = 'RATING_ERROR';
+export const TOKEN = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InRlc3QyQHRlc3QuY29tIiwiZXhwIjoxNTQxNjg1NjUzfQ.jc8Y9cVtQfns4i6g0La0cpXZw0TqRlQ_JziTCM46ToY';
```
This should be an environment variable

koechkevin replied:
this is a token obtained after login so i just used here for the purpose of testing

Arthur236 commented on src/constants/index.js:
```diff
@@ -0,0 +1,5 @@
+export const AVERAGE_RATING = 'AVERAGE_RATING';
+export const RATING_ERROR = 'RATING_ERROR';
+export const TOKEN = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InRlc3QyQHRlc3QuY29tIiwiZXhwIjoxNTQxNjg1NjUzfQ.jc8Y9cVtQfns4i6g0La0cpXZw0TqRlQ_JziTCM46ToY';
+export const BACKEND_URL = 'http://localhost:8000';
```
Same case here

Arthur236 commented:
@koechkevin rebase your branch with develop

Arthur236 commented on src/__test__/components/rating.components.test.jsx:
```diff
@@ -0,0 +1,42 @@
+    const wrapper = mount(<RateArticle store={store} />);
+    expect(wrapper.find(StarRatings).length).toEqual(2);
+  });
+  it('finds only my rating since the articles are undefined', () => {
```
Let's have blank lines to separate each `it` block.
Apply this throughout your file and all other files where this is applicable

koechkevin replied:
Noted. I will rebase, refactor code and implement the changes

koechkevin replied:
Hound raises complains when a space is left after each it block

Arthur236 commented:
Rename this file `src/__test__/components/rating.components.test.jsx` to `src/__test__/components/RatingComponent.test.js`

Arthur236 commented:
We had agreed on the structure to use for our components and containers. Therefore, instead of having:
```bash
|- src
|  |- components
|  |  |- Rating
|  |  |  |- index.js
```

Lets have:
```bash
|- src
|  |- components
|  |  |- Rating
|  |  |  |- RatingComponent.jsx
```

Apply the same to the actions and reducers as well. Have `rating.action.js` and `rating.reducer.js` instead of the folder with the index.js file

Arthur236 commented on package.json:
```diff
@@ -6,19 +6,24 @@
     "history": "^4.7.2",
     "jest": "^23.6.0",
+    "materialize-css": "^1.0.0",
+    "moxios": "^0.4.0",
```
Have you used this anywhere?

koechkevin replied:
For moxios I had used before the sync on tests but I later opted out. I used materialize css to style the mock article.

Arthur236 commented on src/__test__/actions/rating.action.test.js:
```diff
@@ -0,0 +1,116 @@
+  });
+});
+
+describe('failedRating action. It should update the state with the error ', () => {
```
I don't think you should have so many `describe` blocks in one file. Just make sure you give the `it` a descriptive title and it should make sense. Otherwise, you just keep repeating this
```
afterEach(() => {
    store.clearActions();
  });
```

Arthur236 commented on src/__test__/components/RatingComponent.test.js:
```diff
@@ -0,0 +1,40 @@
+import RateArticle from '../../components/Rating/RatingComponent';
+import { rateArticleAction, getRatings } from '../../actions/rating.action';
+
+let wrapper = shallow(<RateArticle slug="hi-there" rating={4} averageRating={3} getRatings={getRatings} rateArticleAction={rateArticleAction} />);
```
Have these props defined in a constant then destructure them when defining your component

Arthur236 commented on src/actions/rating.action.js:
```diff
@@ -0,0 +1,75 @@
+
+const { RATE_ARTICLE, AVERAGE_RATING, RATING_ERROR } = CONSTANTS;
+export const rateArticle = data => ({
+  /**
```
Block comments in javascript are placed above the function.
Refer here `http://usejsdoc.org/about-getting-started.html`

Arthur236 commented on src/api/index.js:
```diff
@@ -1,14 +1,23 @@
 });
 
+const authHeader = () => (TOKEN ? {
+  /**
```
Fix this comment block as well

Arthur236 commented on src/constants/index.js:
```diff
@@ -1,5 +1,9 @@
 };
 
 export default CONSTANTS;
+export const TOKEN = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InRlc3QxQHRlc3QuY29tIiwiZXhwIjoxNTQxODgzNjU2fQ.WvUOntR_I3uV_l0sjmkg-WLA1XErlY6N6guycRS27Ek';
```
Why is the token defined as a constant?
I thought it is generated dynamically every time a user logs in

koechkevin replied:
This should be changed once login has been implemented... I was just using it to test the functionality of this feature

Arthur236 commented on src/mock/article.js:
```diff
@@ -0,0 +1,459 @@
+    title: ['lorem ipsum dolor sit amet'],
+    body: [`
+    Lorem
+    ipsum
```
Why is every word placed on its own line?
This is not ideal at all

koechkevin replied:
let me have this fixed

Arthur236 commented on src/__test__/containers/MyRatingContainer.test.js:
```diff
@@ -0,0 +1,22 @@
+import { mapStateToProps } from '../../containers/Rating/MyRatingContainer';
+
+const state = {
```
Define the `state` inside your describe function and not as a global

Arthur236 commented on src/components/Rating/RatingComponent.jsx:
```diff
@@ -0,0 +1,63 @@
+
+  render() {
+    return (
+      <React.Fragment>
```
The Fragment wrapper is not necessary since you are already wrapping everything in your container div

koechkevin replied:
The container div does not wrap everything

koechkevin committed:
feat(rating): fix

 - fix css styling
- fix  line on token

[Finishes #160577444]

koechkevin committed:
feat(rating): rate an article

- Fix failing tests

[Finishes #160577444]

koechkevin committed:
feat(rating): rate an article

- Add toast message on successful rating of an article

[Finishes #160577444]

koechkevin committed:
feat(rating): rate an article

- adds toast messages after successfully rating an article

[Finishes #160577444]

koechkevin committed:
feat(rating): rate an article

- change the position of rating star and the nature of star as requested by PO

[Finishes #160577444]

koechkevin committed:
fix(rating): rate an article

- remove unecessary condition  in statement

[Finishes #160577444]

Arthur236 commented on src/reducers/initialState.js:
```diff
@@ -10,7 +10,10 @@ export default {
   resetPassword: {
     data: {},
-    errors: {},
+    errors: {} },
```
Have the closing curly brace on its own line

Arthur236 commented:
The PR looks okay. Just one more thing.
If I am the owner of the article, in addition to not being able to see the stars for rating, I shouldnt also be able to see the text highlighted.

<img width="1247" alt="screenshot 2018-11-19 at 20 31 01" src="https://user-images.githubusercontent.com/31339738/48724435-3ff3ca00-ec3a-11e8-91db-76ad8252b0e4.png">

Arthur236 commented:
Kindly rebase as well

koechkevin committed:
fix(rating): rate an article

- ensures user cannot view my ratings component in own article

[Finishes #160577444]

Arthur236 reviewed (approved):
Looks good 👍
